zipkin-local.js

```diff
@@ -0,0 +1,148 @@
+    timeout: 60 * 60 * 60 * 1000000,
+  });
+  const tracer = new _zipkinJavascriptOpentracing.default({
+    localServiceName: SITE_NAME,
+    serviceName: SITE_NAME,
+    sampler: new _zipkin.sampler.CountingSampler(1),
+    traceId128Bit: true,
+    supportsJoin: true,
+    recorder,
+    kind: `client`,
+  });
+  return tracer;
+};
+
+const formatTrace = (trace) => {
+  const { SITE_NAME, traces: {tags} = {}, buildId } = PLUGIN_OPTIONS;
+  trace = JSON.parse(trace);
+  trace.localEndpoint = {};
+  trace.localEndpoint.serviceName = SITE_NAME;
+  if(!trace.tags) {
+    trace.tags = {};
+  }
+  if (trace.annotations) {
+    for (let anno of trace.annotations) {
+      trace.tags[anno.key] = anno.value;
+    }
+  }
+  if (trace.binaryAnnotations) {
+    for (let anno of trace.binaryAnnotations) {
+      trace.tags[anno.key] = anno.value;
+    }
+  }
+  if (trace.name === "run-api") {
+    trace.name += `: ${trace.tags.api}`;
+  }
+  if (trace.name === "run-plugin") {
+    trace.name += `: ${trace.tags.plugin}`;
+  }
+  trace.tags = {...trace.tags, ...tags};
+  trace.tags.thing = 'zipkin-local';
+  trace.tags.buildId = buildId;
+  return JSON.stringify(trace);
+}
+
+const sendTraceQueue = async (queue) => {
+  const { NR_INGEST_KEY } = PLUGIN_OPTIONS;
+  const response = await (0, _nodeFetch.default)(logger.endpoint, {
+    method: `POST`,
+    body: queue,
+    headers: {
+      "Content-Type": "application/json",
+      "Api-Key": NR_INGEST_KEY,
+      "Data-Format": "zipkin",
+      "Data-Format-Version": 2,
+    },
+  });
+
+  if (response.status >= 300) {
+    const err =
+      `[@] gatsby-plugin-newrelic: Unexpected response while sending trace data, status:` +
+      `${response.status}, body: ${postBody}`;
+    if (logger.errorListenerSet) {
+      logger.emit(`error`, new Error(err));
+    }
+    else {
+      console.error(err);
+    }
+  }
+}
+
+const _processQueue = async () => {
+  const { NR_INGEST_KEY, traces: {collectTraces} = {collectTraces: true} } = PLUGIN_OPTIONS;
+  if (!NR_INGEST_KEY || !collectTraces || logger.queue.length <= 0) {
+    return;
+  }
+  try {
+    const formattedQueue = logger.queue.map(formatTrace);
+    const postBody = `[${formattedQueue.join(",")}]`;
+    await sendTraceQueue(postBody);
+  } catch (error) {
+    const err = `[@] gatsby-plugin-newrelic: Error sending trace data ${error}`;
+    if (logger.errorListenerSet) {
+      logger.emit(`error`, new Error(err));
+    }
+    console.error(err);
+  }
+};
+
+/**
+ * Run any tracer cleanup required before the node.js process
+ * exits. For Zipkin HTTP, we must manually process any spans still on
+ * the queue
+ */
+const stop = async () => {
+    // First, write all partial spans to the http logger
+  recorder.partialSpans.forEach((span, id) => {
+    if (span.timeoutTimestamp < now()) {
+      recorder._writeSpan(id);
+    }
+  }); // Then tell http logger to process all spans in its queue
```

Comment:
Does `recorder._writeSpan` happen asynchronously? If so (and we don't await it), we run the risk of a race condition where the partial spans haven't been written to the logger but we start processing the queue anyway.

Reply:
No it actually doesn't so I think we are ok here? https://github.com/openzipkin/zipkin-js/blob/ec89188cf6a07e184ab886c1dfb6c9dc276ddfa4/packages/zipkin/src/batch-recorder.js#L115